specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,143 @@
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Gas Stipend](#gas-stipend)
+- [Limiting Guaranteed Gas](#limiting-guaranteed-gas)
+- [1559 Fee Market](#1559-fee-market)
+- [Rationale for burning L1 Gas](#rationale-for-burning-l1-gas)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+[Deposited transaction](./glossary.md#deposited-transaction) are transactions on L2
+that are initiated on L1. The gas that they use on L2 is bought on L1 via a gas burn
+or a direct payment. We maintain a fee market and hard cap on the amount of gas provided
+to all deposits in a single L1 block.
+
+The gas provided to deposited transactions is sometimes called "guaranteed gas". The
+gas provided to deposited transactions is unique in the regard that it is not
+refundable. It cannot be refunded as it is sometimes paid for with a gas burn and
+there may not be any ETH left to refund.
+
+The **guaranteed gas** is composed of a gas stipend, and of any guaranteed gas the user
+would like to purchase (on L1) on top of that.
+
+## Gas Stipend
+
+Because there is some cost to submitting the transaction and updating the basefee,
+we provide transactions with a small amount of free gas.
+
+If the user requests more `guaranteedGas` than the `gasStipend`, that gas will
+be bought with L1 ETH via a gas burn or by buying it directly. If they request
+less gas than the stipend, they will not be charged.
+
+We provide XX,XXX amount of gas in the stipend. TODO - should this be price
+rather than gas based??
+
+## Limiting Guaranteed Gas
+
+The total amount of guaranteed gas that can be bought in a single L1 block must
+be limited to prevent a denial of service attack against L2 as well as allow the
+total amount of guaranteed gas to be below the L2 block gas limit.
+
+We set limit the total amount of gas buyable via a contract method. It will initially
+be controlled by the Optimism Multisig before being handed over to governance.
+TODO - check that this is the actual plan.
+
+## 1559 Fee Market
+
+To reduce [Priority Gas Auctions](./glossary.md#priority-gas-auction) and accurately price gas, we implement a 1559
+style fee market on L1 with the following pseudocode. We also use this opporunity to
+place a hard limit on the amount of guaranteed gas that is provided.
+
+```text
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 2
+
+curr_basefee: u128, curr_num: u64, curr_bought_gas: u64 = load_and_unpack_storage()
+GUARANTEED_GAS_LIMIT: u64, SANITY_GAS_LIMIT: u64 = load_and_unpack_storage2()
+gas_target = GUARANTEED_GAS_LIMIT // ELASTICITY_MULTIPLIER
+
+# // implies floor division, however because gas_delta is always positive, it is the same as truncating (aka round to 0) division
+# If first deposit of this block, calculate the new basefee and store other info as well.
+if curr_num != block.number {
```

Comment:
I don't know if i follow this because if we do the updates of the base fee inside of this `if curr_num != block.number` then that means only the first deposit of each block can influence the deposit basefee

Reply:
That is correct - the fee is updated once per block at the start of the block. I just realized we may need to adjust this algorithm to handle skipped blocks.

We are intentionally moving away from the design where the fee increases for each tx in the block.